internals-js/src/coreSpec.ts

```diff
@@ -373,7 +373,42 @@ export class CoreSpecDefinition extends FeatureDefinition {
     if (as) {
       args.as = as;
     }
-    schema.schemaDefinition.applyDirective(as ?? this.url.name, args, true);
+
+    // This adds `@link(url: "https://specs.apollo.dev/link/v1.0")` to the "schema" definition. And we have
+    // a choice to add it either the main definition, or to an `extend schema`.
+    //
+    // In theory, always adding it to the main definition should be safe since even if some root operations
+    // can be defined in extensions, you shouldn't have an extension without a definition, and so we should
+    // never be in a case where _all_ root operations are defined in extensions (which would be a problem
+    // for printing the definition itsef since it's syntactically invalid to have a schema definition with
+    // no operations).
+    //
+    // In practice however, graphQL-js has historically accepted extensions without definition for schema,
+    // and we even abuse this a bit with federation out of convenience, so we could end up in the situation
+    // where if we put the directive on the definition, it cannot be printed properly due to the user having
+    // defined all its root operations in an extension.
+    //
+    // We could always add the directive to an extension, and that could kind of work but:
+    // 1. the core/link spec says that the link-to-link application should be the first `@link` of the
+    //   schema, but if user put some `@link` on their schema definition but we always put the link-to-link
+    //   on an extension, then we're kind of not respecting our own spec (in practice, our own code can
+    //   actually handle this as it does not strongly rely on that "it should be the first" rule, but that
+    //   would set a bad example).
+    // 2. earlier versions (pre-#1875) were always putting that directive on the definition, and we wanted
+    //   to avoid suprising users by changing that for not reason.
+    //
+    // So instead, we put the directive on the schema definition unless some extensions exists but no
+    // definition does (that is, no non-extension elements are populated).
+    const schemaDef =  schema.schemaDefinition;
+    // Side-note: this test must be done _before_ we call `applyDirective`, otherwise it would take it into
+    // account.
+    const hasDefinition = schemaDef.hasNonExtensionElements();
+    const directive = schemaDef.applyDirective(as ?? this.url.name, args, true);
```

Comment:
I found this code a bit confusing for a second because of the fact that I was reading `as` as the typescript casting keyword. I'd probably change the variable name for clarity.